lambda/transl_array_comprehension.ml

```diff
@@ -0,0 +1,932 @@
+           overflowed; if it has, we have to double the size of the array.  (The
+           complex case corresponds to [array_size.array_sizing] being
+           [Dynamic_size].)
+       (b) If the array kind is not statically known, we initially created an
+           empty array; we have to check if we're on the first iteration and use
+           the putative first element of the array as the placeholder value for
+           every element of the array.  (The complex case corresponds to
+           [array_kind] being [Pgenarray].)
+       (c) If both (a) and (b) hold, we shouldn't bother checking for an
+           overflowed index on the first loop iteration.
+     The result is that we build the "set the element" behavior in three steps:
+       i.   First, we build the raw "set the element unconditionally" expression
+            ([set_element_raw]).
+       ii.  Then, if necessary, we precede that with the resizing check;
+            otherwise, we leave the raw behavior alone
+            ([set_element_in_bounds]).
+       iii. Then, if necessary, we check to see if we're on the first iteration
+            and create the fresh array instead if so; otherwise, we leave the
+            size-safe behavior alone ([set_element_known_kind_in_bounds]).
+       iv.  Finally, we take the resulting safe element-setting behavior (which
+            could be equal to the result from any of stages i--iii), and follow
+            it up by advancing the index of the element to update.
+  *)
+  let open (val Lambda_utils.int_ops ~loc) in
+  let set_element_raw elt =
+    (* array.(index) <- elt *)
+    Lprim(Parraysetu array_kind, [Lvar array; Lvar index; elt], loc)
+      (* CR aspectorzabusky: Is [array_kind] safe here, since it could be
+         [Pgenarray]?  Do we have to learn which it should be? *)
+  in
+  let set_element_in_bounds elt = match array_sizing with
+    | Fixed_size ->
+        set_element_raw elt
+    | Dynamic_size ->
+        Lsequence(
+          (* Double the size of the array if it's time... *)
+          Lifthenelse(Lvar index < Lvar array_size,
+            lambda_unit,
+            Lsequence(
+              Lassign(array_size, i 2 * Lvar array_size),
+              Lassign(array,      Resizable_array.double ~loc (Lvar array))),
+            Pintval (* [unit] is immediate *)),
+          (* ...and then set the element now that the array is big enough *)
+          set_element_raw elt)
+  in
+  let set_element_known_kind_in_bounds = match array_kind with
+    | Pgenarray ->
+        let is_first_iteration = (Lvar index = l0) in
+        let elt = Ident.create_local "elt" in
+        let make_array = match array_sizing with
+          | Fixed_size ->
+              make_vect ~loc ~length:(Lvar array_size) ~init:(Lvar elt)
+          | Dynamic_size ->
+              Resizable_array.make ~loc Pgenarray (Lvar elt)
+        in
+        (* CR aspectorzabusky: Is Pgenval safe here? *)
```

Comment:
What is `elt` going to be?  If it's a normal OCaml value, as opposed to an unboxed float, then `Pgenval` is correct.

Reply:
`elt` will be an expression the user wrote, which might have type `float` but shouldn't otherwise be magically unboxed, so this should be fine, then.  Would it be better to do any analysis on the type of `body` to pick a better thing than `Pgenval`?